src/main/java/org/wildfly/security/authz/MappedRoleMapper.java

```diff
@@ -0,0 +1,171 @@
+        this.reverseRoleMap = new LinkedHashMap<>();
+        this.removeMapped = removeMapped;
+        this.removeNonMapped = removeNonMapped;
+        this.keyRoles = Roles.fromSet(roleMap.keySet());
+
+        Set<Map.Entry<String, Set<String>>> entrySet = roleMap.entrySet();
+
+        for (Map.Entry<String, Set<String>> entry : entrySet) {
+
+            for (String mappedRole : entry.getValue()) {
+                Set<String> rolesToMappedRole = reverseRoleMap.get(mappedRole);
+
+                if (rolesToMappedRole == null) {
+                    rolesToMappedRole = new LinkedHashSet<>();
+                    reverseRoleMap.put(mappedRole, rolesToMappedRole);
+                }
+
+                rolesToMappedRole.add(entry.getKey());
+            }
+        }
+    }
+
+
+    /**
+     * Custom component method.
+     * Called from WildFly core. Used to include mapped role mapping functionality in older WildFly versions.
+     *
+     * @param configuration map of mapping rules
+     */
+    public void initialize(final Map<String, String> configuration) {
+        configuration.forEach( (key, value) -> {
+            String[] newRoles = value.split(",");
+            for (String newRole : newRoles) {
+                Set<String> rolesToMappedRole = reverseRoleMap.get(newRole);
+                if (rolesToMappedRole == null) {
+                    rolesToMappedRole = new LinkedHashSet<>();
+                    reverseRoleMap.put(newRole, rolesToMappedRole);
+                }
+                rolesToMappedRole.add(value);
+            }
+        } );
+    }
+
+    @Override
+    public Roles mapRoles(Roles rolesToMap) {
+        Roles res = new MappedRoles(rolesToMap, this.reverseRoleMap);
+        if (!removeMapped) {
+            res = res.or(rolesToMap.and(keyRoles));
+        }
+        if (!removeNonMapped) {
+            res = res.or(rolesToMap.minus(keyRoles));
+        }
+        return res;
+    }
+
+    /**
+     * Construct a new {@link Builder} for creating the {@link MappedRoleMapper}.
+     *
+     * @return a new {@link Builder} for creating the {@link MappedRoleMapper}.
+     */
+    public static Builder builder() {
+        return new Builder();
+    }
+
+    /**
+     * A builder for map backed role mappers.
+     */
+    public static class Builder {
+        private Map<String, Set<String>> roleMap;
+        private boolean removeMapped = true;
+        private boolean removeNonMapped = true;
+
+        Builder() {
+        }
+
+        /**
+         * Build and return the resulting {@link MappedRoleMapper}.
+         *
+         * @return the resulting {@link MappedRoleMapper}
+         */
+        public MappedRoleMapper build() {
+            return new MappedRoleMapper(roleMap, removeMapped, removeNonMapped);
+        }
+
+        /**
+         * Set the mapper to remove roles, that were changed by its {@link Map}
```

Comment:
Sounds a bit like the output of the mapping is removed. Maybe would be better something like "to replace original roles by the mapped ones, instead of keeping original and mapped roles abreast" ?